src/extension/infer.rs

```diff
@@ -0,0 +1,855 @@
+                            from_offset: Port::new(Direction::Outgoing, 0),
+                            from_extensions: src_rs,
+                            to: *tgt,
+                            to_offset: Port::new(Direction::Incoming, 0),
+                            to_extensions: tgt_rs,
+                        },
+                    ))
+                }
+            } else {
+                None
+            }
+        } else {
+            None
+        };
+        err.unwrap_or(InferExtensionError::MismatchedConcrete {
+            expected: rs1,
+            actual: rs2,
+        })
+    }
+
+    /// Take a group of equal metas and merge them into a new, single meta.
+    ///
+    /// Returns the set of new metas created and the set of metas that were
+    /// merged.
+    fn merge_equal_metas(&mut self) -> Result<(HashSet<Meta>, HashSet<Meta>), InferExtensionError> {
+        let mut merged: HashSet<Meta> = HashSet::new();
+        let mut new_metas: HashSet<Meta> = HashSet::new();
+        for cc in self.eq_graph.ccs().into_iter() {
+            // Within a connected component everything is equal
+            let combined_meta = self.fresh_meta();
+            for m in cc.iter() {
+                if self.shunted.contains_key(m) {
+                    continue;
+                }
+
+                if let Some(cs) = self.constraints.get(m).cloned() {
+                    for c in cs
+                        .into_iter()
+                        .filter(|c| !matches!(c, Constraint::Equal(_)))
+                    {
+                        self.add_constraint(combined_meta, c.clone());
+                    }
+                    self.constraints.remove(m).unwrap();
+                    merged.insert(*m);
+                    // Record a new meta the first time that we use it; don't
+                    // bother recording a new meta if we don't add any
+                    // constraints. It should be safe to call this multiple times
+                    new_metas.insert(combined_meta);
+                }
+                if let Some(solution) = self.solved.get(m) {
+                    match self.solved.get(&combined_meta) {
+                        Some(existing_solution) => {
+                            if solution != existing_solution {
+                                return Err(self.report_mismatch(
+                                    *m,
+                                    combined_meta,
+                                    solution.clone(),
+                                    existing_solution.clone(),
+                                ));
+                            }
+                        }
+                        None => {
+                            self.solved.insert(combined_meta, solution.clone());
+                        }
+                    }
+                }
+                if self.variables.contains(m) {
+                    self.variables.insert(combined_meta);
```

Comment:
does m need removing from variables?

Reply:
It's harmless to leave it in, but we could probably delete it

Reply:
Deleting it now